### DISCUSSION TITLE
how to run GUI from Linux CLI

### DISCUSSION THREAD

**Discussion by ham-smart** [JaapvanEkris/openrowingmonitor#29] [General]
hi, I just installed app to my raspberry pi and I want to start GUI. So I just tried follow setcap command, restarted my machine, but couldn't open GUI. How can I run GUI? just give me any tips. Thanks.

**Comment by Abasz** [comment]
GUI is accessed from a web browser that is on the same machine via localhost. Port is normally 80 or else if you set it to something else. You can check the logs in syslog if the app actually started ok.

**Comment by JaapvanEkris** [comment]
> hi, I just installed app to my raspberry pi and I want to start GUI. So I just tried follow setcap command, restarted my machine, but couldn't open GUI. How can I run GUI? just give me any tips. Thanks.

What is the version of OpenRowingMonitor you are using? I've updated the install script and libraries yesterday, so it might be something to look at.

What does the GUI does show? Does it start, or is it just the login prompt? And does the webppage Abasz suggested show something?

**Reply by ham-smart** [reply]
Thanks for reply. I tried to install v1beta-updates with new install.sh. And it looks fine, but I couldn't run next command "sudo systemctl status openrowingmonitor" and I couldn't find any openrowingmonitor.service file in my system folder. So I cannot run further. Any tips?

**Reply by JaapvanEkris** [reply]
Did the install procedure complete?

What I notice is that the install script doesn't fail in a very explicit error. Inhad issues with my network, and the npm install or apt-get would time out, and then it just stoped.

Which Raspberry are you installing it on? And what RPI image are you using? Bookworm has switched to a new windowmanager, so I am still looking hiw to install a decent kioskmode on that. Bullseye would be preferred at the moment.

**Reply by ham-smart** [reply]
I reinstalled last update yesterday after I got reply from you.
 
My Raspberry Pi is 4 Model B and installed Raspberry Pi Os Liste 64 bit version on SD card. 
And I removed all in SD card before I tried with last update, in any case. 
I didn't get any timeout error and install log looks ok and I got apt-get update without problem in install log. 

But I got error message `couldn't find any openrowingmonitor.service file` when I runned `sudo systemctl status openrowingmonitor` after install. So I checked /lib/systemd/system folder in root & usr directory and there're no openrowingmonitor.service file. 

That's all what I tried last night :)

**Reply by JaapvanEkris** [reply]
Which OS-version are you using? I noticed that the current Raspberry Manager installs the "Bookworm" version, which still has issues. Could you use the Raspberry Pi manager, the option "Raspberry Pi OS (other)", and then select the "(Legacy 64 Bit) Lite" version.

Addition: I changed the install script to warn users about this. Sorry about any inconvenience caused.

**Reply by ham-smart** [reply]
My current OS is Bullseye, not 100% sure. I can check it again when I back to home :) Thanks for quick reply and update. ;)

**Reply by Techtobi83** [reply]
My setup from 12/2023 to 01/2024 had issues with "Bookworm". You explicit need to choose OS as Japp told you (Bullseye OS Lite legacy 64bit). Only with that setting for me install.sh run full until its installation end. 

If istall script was 100%, normally nothing is to do more, GUI starts automatically after restart OS (also this step is done by install.sh)

**Reply by JaapvanEkris** [reply]
We'll get you there.

I'm currently testing and improving the install script (as I disconnected it from the original repository as they haven't accepted a PR in over a year, despite frequent e-mails asking them to take a decision). I installed ORM a dozen times in the last days, and the process isn't as robust as I hoped.

For me, there are some steps to take in the coming weeks:
* Migrate as many npm packages to a recent version without breaking node V14 (i.e. Pi Zero W) compatibility. This isn't trivial as I suspect the latest Rollup-package kills our webserver, and the latest ESLine packages don't work under Node V14 (seen that issue already). Lars deliberatly locked the Pi Zero to V14 as some other package would break, which I can't test if that is still the case as I have no spare Zero's (the backported versions support up to Node V20).
* Migrate the Raspberry 4 branch to Node.js 18 or 20. The warning of the V16 version is annoying but justified. There is some issue with the gyp-package that requires some flags which aren't supported anymore by newer Node versions. I think I found the new way of setting these flags, but I want to take clear steps here.
* Migrate/support this all to the Bookworm version of  Raspberry OS. As this introduces a changed window manager, I need to find out how this new one can display a webpage without too much overhead.
* Make the install script much more robust against failed installs. On a weaker network connection it can fail to install a critical package and then abort, but as a user installing hundreds of packages people don't see it going bad as apt-get continues installing but the script will crash afterwards.

**Reply by ham-smart** [reply]
Thanks a lot guys. I will try again when I back home and update status asap :)

**Reply by ham-smart** [reply]
Hi, I got GUI finally. Thanks a lot. I can move forward. 

By the way, I wonder if there is any typical setting for Water Rower. 
I tried to change some settings in config file to get the right measure on the screen, but it looks unmatch with my Water Rower.
Any tips?

**Reply by JaapvanEkris** [reply]
I would start at the settings of the Sportstech WRX700, which is a water based rower

**Reply by Raymond-ArrA** [reply]
I have tried with no luck, to change " rowerSettings: rowerProfiles.DEFAULT, " to Sportstech WRX700 in config.js, because I have almost the same kind of rower. Even tried to create one duplicated set of settings with another name in rowerProfiles.js, 

Only if i change the data in the actual DEFAULT set of settings in rowerProfiles.js, the changes is made. This is easy to see, based on the Drag Factor in the WebGui, after restarting the device. it seems to be stuck on 1500. Either Default or Fallback to Default is chosen. 

Is there any bug that prevent me from changing from default to another rower?

**Reply by Abasz** [reply]
Can you please send your config.js file here? Its probably a ill formed settings file (missing a curly or an extra comment etc.)

**Reply by JaapvanEkris** [reply]
Small practical tip: look for the comma's. In the config file, the last line of each structure should NOET contain a comma.

**Reply by Raymond-ArrA** [reply]
ah, I have no access to the actual file right now, but in short, I started from scratch and with the default config.js file, located in config/config.js - I removed the "//" in front of the setting (just to test) :
 
From this:
// EXAMPLE ROWER CONFIG : using a DKN R-320 Air Rower as is
// rowerSettings: rowerProfiles.DKN_R320

To this:
// EXAMPLE ROWER CONFIG : using a DKN R-320 Air Rower as is
    rowerSettings: rowerProfiles.DKN_R320
  
Not even this works. so I tried to replicate the line and change the rower profile.

  // EXAMPLE ROWER CONFIG : using a DKN R-320 Air Rower as is
  // rowerSettings: rowerProfiles.DKN_R320
      rowerSettings: rowerProfiles.Sportstech_WRX700

But had no luck. And it seems to fallback to Default anyways.
I tried to change the fallback in default.config.js - just to try, and that ended up with no web-browser startup :p

So.. only small changes from the original config.js file.
When this did not work, I tride to change values under config/rowerProfiles.js - the same way. I left the original with "//" in front, and replicated the setting under, like with the rowerSettings over - what did the trick. Even tho it is a bit of tweaking, I can work from here to test it, But I really want to create a new profile with the TrainX WaterRower pro, instead of changing DEFAULTS, but I cant get it to work

I can always get back to the original working model, but not change the profile setting.
So.. can you try to change your profile to another and actually check if it change the profile?

**Reply by Abasz** [reply]
I am pretty sure that your config file is incorrectly formed and that is why it does not get picked up. We need to hole file to verify this, not just samples, but the hole file. I am like 99% sure that you are missing a coma at the end as it is not in the commented out line in the example.

**Reply by Abasz** [reply]
```js
'use strict'
/*
  Open Rowing Monitor, https://github.com/laberning/openrowingmonitor

  You can modify this file to configure Open Rowing Monitor to your needs.
  This file should be placed in the 'config' folder of Open Rowing Monitor.

  All available configuration parameters are visible in config/config.default.js
  To modify a parameter, copy it to this file and modify the value.

  Changes to this file are persisted when you update to new versions.
*/
// eslint-disable-next-line no-unused-vars
import rowerProfiles from './rowerProfiles.js'

export default {
    // example: change the default log level:
    loglevel: {
      default: 'debug'
    },

  // The rower specific settings. Either choose a profile from config/rowerProfiles.js (see
  // https://github.com/laberning/openrowingmonitor/blob/main/docs/Supported_Rowers.md) or define
  // the settings manually (see https://github.com/laberning/openrowingmonitor/blob/main/docs/rower_settings.md
  // on how to do this). If you find good settings for a new rowing device please send them to us (together
  // with a raw recording of at least 10 strokes) so we can add the device to the profiles.

  // EXAMPLE ROWER CONFIG : using a DKN R-320 Air Rower as is
  rowerSettings: rowerProfiles.Sportstech_WRX700, // NOTE THE COMA HERE AT THE END OF THE LINE
}
```

**Reply by JaapvanEkris** [reply]
I'm slowly convinced we need to add a name to the rower profiles and have the config manager report what is loaded. The current merge process fails unnoticed and it sucks.

**Comment by JaapvanEkris** [comment]
```
'use strict'
/*
  Open Rowing Monitor, https://github.com/laberning/openrowingmonitor

  You can modify this file to configure Open Rowing Monitor to your needs.
  This file should be placed in the 'config' folder of Open Rowing Monitor.

  All available configuration parameters are visible in config/config.default.js
  To modify a parameter, copy it to this file and modify the value.

  Changes to this file are persisted when you update to new versions.
*/
// eslint-disable-next-line no-unused-vars
import rowerProfiles from './rowerProfiles.js'

export default {
    // example: change the default log level:
    loglevel: {
      default: 'debug'
    },

  // The rower specific settings. Either choose a profile from config/rowerProfiles.js (see
  // https://github.com/laberning/openrowingmonitor/blob/main/docs/Supported_Rowers.md) or define
  // the settings manually (see https://github.com/laberning/openrowingmonitor/blob/main/docs/rower_settings.md
  // on how to do this). If you find good settings for a new rowing device please send them to us (together
  // with a raw recording of at least 10 strokes) so we can add the device to the profiles.

  // EXAMPLE ROWER CONFIG : using a DKN R-320 Air Rower as is
  rowerSettings: rowerProfiles.Concept2_RowErg

  // EXAMPLE ROWER CONFIG: Just set custom rower settings to make it work
  // rowerSettings: {
  //   numOfImpulsesPerRevolution: 1,
  //   dragFactor: 0.03,
  //   flywheelInertia: 0.3
  // }

  // EXAMPLE ROWER CONFIG: set a rower profile, but overwrite some settings:
  // rowerSettings: Object.assign(rowerProfiles.DKN_R320, {
  //   autoAdjustDragFactor: true
  // })
}
```
Similar to @Abasz , but his solution contains objects that are not in the release branch yet.

**Reply by Abasz** [reply]
> Similar to @Abasz , but his solution contains objects that are not in the release branch yet.

You are right sorry, I corrected it, mine should be correct now

**Reply by JaapvanEkris** [reply]
> You are right sorry, I corrected it, mine should be correct now

When you remove the comma from the last line after Sportstech_WRX700, it will parse (I absolutely hate this JSON-based approach, as comma's and their placement suck, but that's JavaScript for you...). Had a similar issue with the package.json today...

**Reply by Abasz** [reply]
I don't think the comma is a Syntax error there. Trailing commas are allowed in the syntax for JS objects

**Reply by Raymond-ArrA** [reply]
I am going to try this later and get back to you. Guess I'll Just go for a fresh install, both OS and opemrowingmonitor, to ensure that there is no faults from start

**Reply by Raymond-ArrA** [reply]
Finally.. I dont know what was the Deal. Fresh install os, fresh install openrowingmonitor, and then Just wrote a New line with rowerSettings: ...

Now I got the wrx700 profile, ant this really seems to be more of a something like my TrainX WaterRower pro machine.

None of these rowers are correct, but one computer told me I was olympic master, and one told me I was a noobish 90 year old. Differences in 10 kmh takes the breath of everyone 😅

now my 10 Pulls reporter 9, and time, distance, etc were actually looking like something. But.

10 Pulls (says 9)
52 meters
3:38/500m
Drive 0.66m
Drive 0.95 sec
Recovery 1.24 sec
27 strokes pr. Min

These looks something like ok. 
But watt is calculate to 34 w.. 
Is that correct? 🤔 Maybe I am a noobish 90 y/o 😅
But.. Is there some values to correct it, If it is wrong?

**Reply by Abasz** [reply]
watt will depend on the inertia and drag, that is probably completely different-. Please have a look at the readme it explains this. There is a discussion on how to measure inertia as well.

**Reply by Raymond-ArrA** [reply]
Okay, thanks. I will read more 😅. I love this project. The only rowing monitor that May complete with concept2 that I have found.. Mostly because I need to tweak these options.. Thanks a lot 🙏

**Comment by Raymond-ArrA** [comment]
The dragFactor on the WaterRower is set to 32000.. Can you explain the values accepted, when default is 1500? The gap between looks dramatically high, though I really still thinknthere is something that is not right.. Could it be that 32000 is a number too high, so the system falls back to another value? I don't know, but I really dont get it.. something has obviousely changed, but still I am rowing on a watt level between 20 and 40 watt.. And it takes me almost 50 strokes to get 200 meters.. What doesnt sound right to me 😅

**Reply by Raymond-ArrA** [reply]
It might also be inertia.. 🤔  But looking to understand them both 😅

**Reply by Gordon-Shumway2** [reply]
Dragfactors of around 30000 are normal for water-based rowers. My FDF neon pro V with a water tank volume of 17 liter has a drag factor of 31500 and an intertia of 0.74.

It might help if you 
- stated which rower you are actually using (you mentioned  "TrainX WaterRower pro"  but googling that returns no hits) 
- uploaded your config.js and rowerprofiles.js
- would turn on debugging, 
- add/modify "autoAdjustDragFactor: true,", restart the ORM service, do ten to twenty strokes with some effort (but not full force) and upload the .tcx file for that session.

Cheers,
Michael

**Reply by Raymond-ArrA** [reply]
Right now I dont even find it myself. Look up masterfit w rower 400. Identical rower as far as I can see.

I am trying your way as soon as I can.. 
![20240211_192708](https://github.com/JaapvanEkris/openrowingmonitor/assets/120331200/945adbe9-c1ab-4dd3-8c5d-3b8493ec5c68)
